code/modules/reagents/chemistry/equilibrium.dm

```diff
@@ -0,0 +1,326 @@
+		else if (cached_temp >= reaction.optimal_temp)
+			delta_t = 1
+		else //too hot
+			delta_t = 0
+			to_delete = TRUE
+			return
+	else
+		if (cached_temp > reaction.optimal_temp && cached_temp <= reaction.required_temp)
+			delta_t = (((cached_temp - reaction.required_temp)**reaction.temp_exponent_factor)/((reaction.optimal_temp - reaction.required_temp)**reaction.temp_exponent_factor))
+		else if (cached_temp <= reaction.optimal_temp)
+			delta_t = 1
+		else //Too cold
+			delta_t = 0
+			to_delete = TRUE
+			return
+	//Catalyst modifier
+	delta_t *= thermic_mod
+
+	purity = delta_ph//set purity equal to pH offset
+
+	//Then adjust purity of result with beaker reagent purity. 
+	purity *= reactant_purity(reaction)
+
+	//Then adjust it from the input modifier
+	purity *= purity_modifier
+
+	//Now we calculate how much to add - this is normalised to the rate up limiter
+	var/delta_chem_factor = (reaction.rate_up_lim*delta_t)*delta_time//add/remove factor
+	var/total_step_added = 0
+	//keep limited
+	if(delta_chem_factor > step_target_vol)
+		delta_chem_factor = step_target_vol
+	else if (delta_chem_factor < CHEMICAL_VOLUME_MINIMUM)
+		delta_chem_factor = CHEMICAL_VOLUME_MINIMUM
+	//Normalise to multiproducts
+	delta_chem_factor /= product_ratio
+	//delta_chem_factor = round(delta_chem_factor, CHEMICAL_QUANTISATION_LEVEL) // Might not be needed - left here incase testmerge shows that it does. Remove before full commit.
+
+	//Call any special reaction steps BEFORE addition
+	reaction.reaction_step(src, holder, delta_chem_factor, purity)
+
+	//Calculate how much product to make and how much reactant to remove factors..
+	for(var/reagent in reaction.required_reagents)
+		holder.remove_reagent(reagent, (delta_chem_factor * reaction.required_reagents[reagent]), safety = TRUE)
+		//Apply pH changes
+		holder.adjust_specific_reagent_ph(reagent, (delta_chem_factor * reaction.required_reagents[reagent])*(reaction.H_ion_release*h_ion_mod))
+
+	for(var/product in reaction.results)
+		//create the products
+		var/step_add = delta_chem_factor * reaction.results[product]
+		holder.add_reagent(product, step_add, null, cached_temp, purity, override_base_ph = TRUE)
+		//Apply pH changes
+		holder.adjust_specific_reagent_ph(product, step_add*reaction.H_ion_release)
+		reacted_vol += step_add
+		total_step_added += step_add
+
+	#ifdef TESTING //Kept in so that people who want to write fermireactions can contact me with this log so I can help them
+	debug_world("Reaction vars: PreReacted:[reacted_vol] of [step_target_vol]. delta_t [delta_t], multiplier [multiplier], delta_chem_factor [delta_chem_factor] Pfactor [product_ratio], purity of [purity] from a delta_ph of [delta_ph]. DeltaTime: [delta_time]")
+	#endif
+		
+	//Apply thermal output of reaction to beaker
+	holder.chem_temp = round(cached_temp + ((reaction.thermic_constant* total_step_added)*thermic_mod))
```

Comment:
`thermic_constant` and other instances of `chem_temp` being modified should utilize `adjust_thermal_energy()` so you can have the reaction sit with a bunch of water to soak up the heat rather than the heat increase of the mixer not be influenced by non-reacting reagents

Reply:
I put this is, but as a result how heat works in reactions will be fundementally different now. I kept in the old option as a reaction flag - mostly because it would trivialise the meth recipie I put in place.